File: kotlin/diffie-hellman/src/main/kotlin/DiffieHellman.kt
import java.math.BigInteger
import kotlin.random.Random

object DiffieHellman {

    private val TWO = BigInteger.valueOf(2)

    fun privateKey(prime: BigInteger): BigInteger {
        require(prime > TWO)

        while (true) {
            val candidate = Random.nextLong(2, prime.toLong()).toBigInteger()

            if (candidate >= TWO && candidate < prime) {
                return candidate
            }
        }
    }

    fun publicKey(p: BigInteger, g: BigInteger, privKey: BigInteger): BigInteger {
        return g.modPow(privKey, p)
    }

    fun secret(prime: BigInteger, publicKey: BigInteger, privateKey: BigInteger): BigInteger {
        return publicKey.modPow(privateKey, prime)
    }
}

File: csharp/diffie-hellman/DiffieHellman.cs
using System.Numerics;

public static class DiffieHellman
{
    private static readonly BigInteger TWO = new(2);

    public static BigInteger PrivateKey(BigInteger primeP) 
    {
        var rnd = new Random();
        while (true)
        {
            BigInteger candidate = rnd.NextInt64(2, (long)primeP);

            if (candidate >= TWO && candidate < primeP)
            {
                return candidate;
            }
        }
    }

    public static BigInteger PublicKey(BigInteger primeP, BigInteger primeG, BigInteger privateKey) => BigInteger.ModPow(primeG, privateKey, primeP);

    public static BigInteger Secret(BigInteger primeP, BigInteger publicKey, BigInteger privateKey) => BigInteger.ModPow(publicKey, privateKey, primeP);
}